Mindbox.Data.Linq/Mapping/AttributedMetaModel.cs

```diff
@@ -204,21 +158,23 @@ internal MetaTable GetTableNoLocks(Type rowType)
 			var attrs = GetTableAttributes(root, true);
 			if (!attrs.Any())
 			{
-				metaTables.Add(rowType, null);
+				metaTables.TryAdd(rowType, null);
 				return null;
 			}
 
-			if (!metaTables.TryGetValue(root, out table))
+			table = metaTables.GetOrAdd(root, type =>
 			{
-				table = new AttributedMetaTable(this, attrs.First(), root);
-				foreach (var inheritanceType in table.RowType.InheritanceTypes)
-					metaTables.Add(inheritanceType.Type, table);
-			}
+				var aTable = new AttributedMetaTable(this, attrs.First(), root);
+				foreach (var inheritanceType in aTable.RowType.InheritanceTypes)
+					metaTables.TryAdd(inheritanceType.Type, aTable);
```

Comment:
It's possible here that parallel threads will register inheritance type meta tables referencing different instances of the root type meta table

Reply:
Why do you think it could be a problem? They have the same root anyway, so all of them will be registered

Reply:
It's hard to tell right away if that will be a problem or not. It can be, for example, that some code relies on reference equality which will be violated.

I'm not sure whether it's critical to register derived types at the same time as the root type. If not, we could just skip it. Or we could do it after `GetOrAdd` and only when it actually added a new item.

Reply:
To illustrate the problematic scenario, let's imagine that we have types A, B: A and two threads are trying to get the meta type for A in parallel:

1. `Thread #1` tries to get the meta info from the dictionary for type A and it's not there
1. `Thread #1` creates object MetaA1 for type A
1. `Thread #2` tries to get the meta info from the dictionary for type A and it's not there
1. `Thread #2` creates object MetaA2 for type A
1. `Thread #2` creates object MetaB2 for type B, MetaB2 references MetaA2
1. `Thread #2` adds object MetaB2 to the dictionary
1. `Thread #1` creates object MetaB1 for type B, MetaB1 references MetaA1
1. `Thread #1` fails to add object MetaB1 to the dictionary because object MetaB2 is already added
1. `Thread #1` adds object MetaA1 to the dictionary
1. `Thread #2` fails to add object MetaA2 to the dictionary because object MetaA1 is already added

As the result, the dictionary contains object MetaA1 and object MetaB2 which references object MetaA2 instead of MetaA1